src/ProductConstructionService/ProductConstructionService.Api/Controllers/GitHubWebhookEventProcessor.cs

```diff
@@ -0,0 +1,132 @@
+
+    public GitHubWebhookEventProcessor(ILogger<GitHubWebhookEventProcessor> logger, IGitHubTokenProvider tokenProvider, BuildAssetRegistryContext context)
+    {
+        _logger = logger;
+        _tokenProvider = tokenProvider;
+        _context = context;
+    }
+
+    protected override async Task ProcessInstallationWebhookAsync(WebhookHeaders headers,
+        InstallationEvent installationEvent,
+        InstallationAction action)
+    {
+        switch(installationEvent.Action)
+        {
+            case InstallationActionValue.Deleted:
+                await RemoveInstallationRepositoriesAsync(installationEvent.Installation.Id);
+                break;
+            case InstallationActionValue.Created:
+                await SynchronizeInstallationRepositoriesAsync(installationEvent.Installation.Id);
+                break;
+            default:
+                _logger.LogError("Received unknown installation action `{action}`", installationEvent.Action);
+                break;
+        }
+    }
+
+    protected override async Task ProcessInstallationRepositoriesWebhookAsync(WebhookHeaders headers, InstallationRepositoriesEvent installationRepositoriesEvent, InstallationRepositoriesAction action)
+    {
+        await SynchronizeInstallationRepositoriesAsync(installationRepositoriesEvent.Installation.Id);
+    }
+
+    private async Task RemoveInstallationRepositoriesAsync(long installationId)
+    {
+        foreach (Maestro.Data.Models.Repository repo in await _context.Repositories.Where(r => r.InstallationId == installationId).ToListAsync())
+        {
+            _context.Update(repo);
+            repo.InstallationId = 0;
+        }
+
+        await _context.SaveChangesAsync();
+    }
+
+    private async Task SynchronizeInstallationRepositoriesAsync(long installationId)
+    {
+        var token = await _tokenProvider.GetTokenForInstallationAsync(installationId);
+        IReadOnlyList<Repository> gitHubRepos = await GetAllInstallationRepositories(token);
+
+        List<Maestro.Data.Models.Repository> toRemove =
+            await _context.Repositories.Where(r => r.InstallationId == installationId).ToListAsync();
+
+        foreach (Repository repo in gitHubRepos)
+        {
+            Maestro.Data.Models.Repository? existing = await _context.Repositories.FindAsync(repo.HtmlUrl);
+            if (existing == null)
+            {
+                _context.Repositories.Add(
+                    new Maestro.Data.Models.Repository
+                    {
+                        RepositoryName = repo.HtmlUrl,
+                        InstallationId = installationId
+                    });
+            }
+            else
+            {
+                toRemove.Remove(existing);
+                existing.InstallationId = installationId;
+                _context.Update(existing);
+            }
+        }
+
+        foreach (Maestro.Data.Models.Repository repository in toRemove)
```

Comment:
Would it be maybe more readable if we just created the `toRemove` here with a LINQ query? Something like `gitHubRepos.Except(...)`? It took me some time to understand why we remove from toRemove